src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs

```diff
@@ -82,153 +90,198 @@ public static bool IsLockFileStillValid(DependencyGraphSpec dgSpec, PackagesLock
-                            StringComparer.InvariantCultureIgnoreCase,
-                            StringComparer.InvariantCultureIgnoreCase))
-            {
-                return false;
-            }
+                    (var hasProjectDependencyChanged, var pmessage) = HasProjectDependencyChanged(framework.Dependencies, directDependencies, target.TargetFramework);
+                    if (hasProjectDependencyChanged)
+                    {
+                        // lock file is out of sync
+                        invalidReasons.Add(pmessage);
+                    }
 
-            // Validate all P2P references
-            foreach (var framework in project.RestoreMetadata.TargetFrameworks)
-            {
-                var target = nuGetLockFile.Targets.FirstOrDefault(
-                    t => EqualityUtility.EqualsWithNullCheck(t.TargetFramework, framework.FrameworkName));
+                    var transitiveDependenciesEnforcedByCentralVersions = target.Dependencies.Where(dep => dep.Type == PackageDependencyType.CentralTransitive).ToList();
+                    var transitiveDependencies = target.Dependencies.Where(dep => dep.Type == PackageDependencyType.Transitive).ToList();
 
-                if (target == null)
-                {
-                    // a new target found in the dgSpec so invalidate existing lock file.
-                    return false;
+                    (var hasTransitiveDependencyChanged, var tmessage) = HasProjectTransitiveDependencyChanged(framework.CentralPackageVersions, transitiveDependenciesEnforcedByCentralVersions, transitiveDependencies);
+                    if (hasTransitiveDependencyChanged)
+                    {
+                        // lock file is out of sync
+                        invalidReasons.Add(tmessage);
+                    }
                 }
 
-                var queue = new Queue<Tuple<string, string>>();
-                var visitedP2PReference = new HashSet<string>();
-
-                foreach (var projectReference in framework.ProjectReferences)
+                // Validate all P2P references
+                foreach (var framework in project.RestoreMetadata.TargetFrameworks)
                 {
-                    if (visitedP2PReference.Add(projectReference.ProjectUniqueName))
+                    var target = nuGetLockFile.Targets.FirstOrDefault(
+                        t => EqualityUtility.EqualsWithNullCheck(t.TargetFramework, framework.FrameworkName));
+
+                    if (target == null)
                     {
-                        var spec = dgSpec.GetProjectSpec(projectReference.ProjectUniqueName);
-                        queue.Enqueue(new Tuple<string, string>(spec.Name, projectReference.ProjectUniqueName));
+                        // This should never be hit. A hit implies that project.RestoreMetadata.TargetsFrameworks and project.TargetsFrameworks are not the same.
+                        invalidReasons.Add((string.Format(
```

Comment:
This is not really a new framework. 
Something is broken.

The PackageSpec model has TargetFrameworkInformation and ProjectRestoreFrameworkInfo, and both should be equivalent. 

If anything, this should be more of a `file a bug` type of message.

Reply:
Changed this case to exception with 'file a bug` type of message.